Use email as notification sink for newly created notifications

ean opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Commit:
Use email as notification sink for newly created notifications

mpolden commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/stubs/MockMailer.java:
```diff
@@ -33,7 +33,13 @@ public String domain() {
+        if (!mails.containsKey(recipient)) {
+            return List.of();
+        }
         return mails.get(recipient);
```
Can be simplifed to `mails.getOrDefault(recipient, List.of())`.

mpolden commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/stubs/MockMailer.java:
```diff
@@ -33,7 +33,13 @@ public String domain() {
     }
 
+    public void discard() {
+        mails.keySet().forEach(k -> mails.remove(k));
```
Can be simplifed to `mails.clear()`. Consider naming the method `reset`.

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notify/Notifier.java:
```diff
@@ -0,0 +1,68 @@
+    private static final Logger log = Logger.getLogger(Notifier.class.getName());
+
+    public Notifier(CuratorDb curatorDb, Mailer mailer) {
+        this.curatorDb = curatorDb;
+        this.mailer = mailer;
```
Nit: Assign to result of `Objects.requireNonNull`.

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notify/Notifier.java:
```diff
@@ -0,0 +1,68 @@
+                dispatch(notification, contacts.stream().map(c -> (TenantContacts.EmailContact) c).collect(Collectors.toList()));
+                break;
+            default:
+                log.fine("Unknown TenantContact Type: " + type);
```
This should throw instead so that sending notifications for unknown contact types doesn't fail silently.

ean replied:
I was a bit back and forth on that. Throwing here will also block a notification being persisted to zookeeper. Is that OK? We can alternatively move the dispatch to after the ZK transaction has been committed. But that again can lead to lost mails..

mpolden replied:
Sending can fail for other reasons too, i.e. the mailer service being unavailable. So it's better that sending the notification happens after the write and outside the write lock. Log an error if it fails.

If lost emails becomes a problem we should create a queue for them, and a separate maintainer that consumes it (and retries on failures).

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notify/Notifier.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private Mail mailOf(Notification n, Collection<String> recipients) {
+        var subject = new Formatter().format("[%s] Vespa Notification for %s", n.level().toString().toUpperCase(), n.type().name());
```
Any reason to not use `Text.format`?

ean replied:
Me not knowing about it. Changed.

Commit:
Add blackhole support to MockMailer

This will ignore all sent mail to avoid storing it in memory where that is not required.
